services/streammanager/googlecloudfunction/googlecloudfunction.go

```diff
@@ -0,0 +1,226 @@
+package cloudfunctions
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/http"
+	"regexp"
+	"strings"
+
+	"google.golang.org/api/cloudfunctions/v1"
+	"google.golang.org/api/googleapi"
+	"google.golang.org/api/idtoken"
+
+	"google.golang.org/api/option"
+
+	"github.com/rudderlabs/rudder-go-kit/logger"
+	backendconfig "github.com/rudderlabs/rudder-server/backend-config"
+	"github.com/rudderlabs/rudder-server/services/streammanager/common"
+	"github.com/tidwall/gjson"
+)
+
+type Config struct {
+	Credentials            string `json:"credentials"`
+	FunctionEnvironment    string `json:"functionEnvironment"`
+	RequireAuthentication  bool   `json:"requireAuthentication"`
+	GoogleCloudFunctionUrl string `json:"googleCloudFunctionUrl"`
+}
+
+type Client struct {
+	service *cloudfunctions.Service
+	opts    common.Opts
+}
+
+var pkgLogger logger.Logger
+
+func init() {
+	pkgLogger = logger.NewLogger().Child("streammanager").Child("GoogleCloudFunction")
+}
+
+type GoogleCloudFunctionProducer struct {
+	client *Client
+	config *Config
+}
+
+// NewProducer creates a producer based on destination config
+func NewProducer(destination *backendconfig.DestinationT, o common.Opts) (*GoogleCloudFunctionProducer, error) {
+	var config Config
+	jsonConfig, err := json.Marshal(destination.Config)
+	if err != nil {
+		return nil, fmt.Errorf("[GoogleCloudFunction] Error while marshalling destination config :: %w", err)
+	}
+	err = json.Unmarshal(jsonConfig, &config)
+	if err != nil {
+		return nil, fmt.Errorf("[GoogleCloudFunction] error  :: error in GoogleCloudFunction while unmarshalling destination config:: %w", err)
+	}
+
+	opts := []option.ClientOption{
+		option.WithCredentialsJSON([]byte(config.Credentials)),
+	}
+
+	service, err := generateService(destination, opts...)
+	// If err is not nil then retrun
+	if err != nil {
+		pkgLogger.Errorf("[GoogleCloudFunction] error  :: %w", err)
+		return nil, err
+	}
+	client := &Client{service, o}
+	destConfig := &Config{config.Credentials, config.FunctionEnvironment, config.RequireAuthentication, config.GoogleCloudFunctionUrl}
+
+	return &GoogleCloudFunctionProducer{client, destConfig}, err
+}
+
+func (producer *GoogleCloudFunctionProducer) Produce(jsonData json.RawMessage, _ interface{}) (statusCode int, respStatus, responseMessage string) {
+	destConfig := producer.config
+	parsedJSON := gjson.ParseBytes(jsonData)
+
+	if destConfig.FunctionEnvironment == "gen1" && destConfig.RequireAuthentication {
```

Comment:
what happens if env is gen1 and RequireAuth is false?

Reply:
It will make REST API call similar to gen2 function as to call through SDK we need to create a client which requires credentials.